File: csharp/1368.minimum-cost-to-make-at-least-one-valid-path-in-a-grid.cs
/*
 * @lc app=leetcode id=1368 lang=csharp
 *
 * [1368] Minimum Cost to Make at Least One Valid Path in a Grid
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static (int DR, int DC)[] Dirs = new (int DR, int DC)[] {
        (DR: 0, DC: 1),
        (DR: 0, DC: -1),
        (DR: 1, DC: 0),
        (DR: -1, DC: 0),
    };

    private static void MinCost(int m, int n, int x, int y, int c, int[][] grid, int[,] cost, Queue<(int R, int C)> queue) {
        while (x >= 0 && x < m && y >= 0 && y < n && cost[x, y] == 0) {
            cost[x, y] = c;
            var dir = Dirs[grid[x][y] - 1];
            queue.Enqueue((R: x, C: y));
            x += dir.DR;
            y += dir.DC;
        }
    }

    public int MinCost(int[][] grid) {
        int m = grid.Length, n = grid[0].Length, c = 1;
        int[,] cost = new int[m, n];
        var queue = new Queue<(int R, int C)>();
        MinCost(m, n, 0, 0, c, grid, cost, queue);
        while (queue.Count > 0 && cost[m - 1, n - 1] == 0) {
            ++c;
            for (int i = queue.Count; i > 0; --i) {
                var curr = queue.Dequeue();
                foreach (var dir in Dirs) {
                    MinCost(m, n, curr.R + dir.DR, curr.C + dir.DC, c,  grid, cost, queue);
                }
            }
        }
        return cost[m - 1, n - 1] - 1;
    }
}
// @lc code=end

File: java/1368.minimum-cost-to-make-at-least-one-valid-path-in-a-grid.java
/*
 * @lc app=leetcode id=1368 lang=java
 *
 * [1368] Minimum Cost to Make at Least One Valid Path in a Grid
 */

import java.util.ArrayDeque;
import java.util.Queue;

// @lc code=start
class Solution {

    private static final Pair[] DIRS = new Pair[] {
        new Pair(0, 1),
        new Pair(0, -1),
        new Pair(1, 0),
        new Pair(-1, 0),
    };

    public int minCost(int[][] grid) {
        int m = grid.length, n = grid[0].length, cost = 1;
        int[][] costs = new int[m][n];
        Queue<Pair> queue = new ArrayDeque<>();
        add(grid, 0, 0, 1, costs, queue);
        while (costs[m - 1][n - 1] == 0) {
            ++cost;
            for (int i = queue.size(); i > 0; --i) {
                Pair curr = queue.poll();
                for (Pair dir : DIRS) {
                    add(grid, curr.r() + dir.r(), curr.c() + dir.c(), cost, costs, queue);
                }
            }
        }
        return costs[m - 1][n - 1] - 1;
    }

    private static void add(int[][] grid, int r, int c, int cost, int[][] costs, Queue<Pair> queue) {
        int m = grid.length, n = grid[0].length;
        while (r >= 0 && r < m && c >= 0 && c < n && costs[r][c] == 0) {
            costs[r][c] = cost;
            queue.offer(new Pair(r, c));
            Pair dir = DIRS[grid[r][c] - 1];
            r = dir.r() + r;
            c = dir.c() + c;
        }
    }

    private record Pair(int r, int c) {}
}
// @lc code=end

